run_matrix.py

```diff
@@ -0,0 +1,161 @@
+    return test_path / test_filename
+
+
+def _get_fixture(charm_config: _CharmTestConfig, charm_path: Path):
+    """Get the tester fixture from a charm."""
+    fixture_path = charm_path / FIXTURE_PATH
+    fixture_id = FIXTURE_IDENTIFIER
+    if charm_config.test_setup:
+        if charm_config.test_setup["location"]:
+            fixture_path = charm_path / Path(charm_config.test_setup["location"])
+        if charm_config.test_setup["identifier"]:
+            fixture_id = charm_config.test_setup["identifier"]
+    return fixture_path, fixture_id
+
+
+def _setup_venv(charm_path: Path) -> Path:
+    """Create the venv for a charm and return the path to its python."""
+    logging.info(f"Installing dependencies in venv for {charm_path}")
+    original_wd = os.getcwd()
+    os.chdir(charm_path)
+    # Create the venv and install the requirements
+    try:
+        subprocess.check_call("python -m venv ./.interface-venv", shell=True, stdout=subprocess.DEVNULL)
+        subprocess.check_call(
+            ".interface-venv/bin/python -m pip install setuptools pytest git+https://github.com/canonical/interface-tester-pytest@main",
+            shell=True,
+            stdout=subprocess.DEVNULL,
+            stderr=subprocess.DEVNULL
+        )
+        subprocess.check_call(
+            ".interface-venv/bin/python -m pip install -r requirements.txt",
+            shell=True,
+            stdout=subprocess.DEVNULL,
+            stderr=subprocess.DEVNULL
+        )
+    except:
+        raise SetupError("venv setup failed")
+    os.chdir(original_wd)
+    return charm_path / ".interface-venv/bin/python"
+
+
+def test_charm(charm_path: Path, test_path: Path):
+    """Run the interface test for a given charm and interface."""
+    logging.info(f"Running tests for {charm_path}")
+    original_wd = os.getcwd()
+    os.chdir(charm_path)
+    try:
+        subprocess.check_call(
+            f"PYTHONPATH=src:lib .interface-venv/bin/python -m pytest {test_path}",
+            shell=True
+        )
+    except:
+        raise InterfaceTestError
+    os.chdir(original_wd)
+
+
+def run_interface_tests(path: Path, include: str = "*", clean=True):
+    """Run the tests for the specified interfaces, defaulting to all."""
+    if clean:
+        _clean()
+    test_results = {}
+    for interface, x in collect_tests(path=path, include=include).items():
+        test_results[interface] = {}
+        logging.info(f"Running tests for interface: {interface}")
+        for _, y in x.items():
+            for role in ["provider", "requirer"]:
+                test_results[interface][role] = {}
+                for charm_config in y[role]["charms"]:
+                    last_result = "success"
+                    logging.info(f"Charm: {charm_config.name}")
+                    try:
+                        charm_path, test_path = prepare_repo(charm_config, interface)
+                        test_charm(charm_path, test_path)
+                    except:
```

Comment:
```python
                    try:
                         charm_path, test_path = prepare_repo(charm_config, interface)
                    except:
                         # log: setup has failed; exc_info=True
                         continue
                    try:
                         test_charm(charm_path, test_path)
                    except:
                         # log: interface tests failed; exc_info=True
                         continue
```

Reply:
This doesn't work correctly because we set the test result for the output json after that block, so we can't `continue`; but I can catch the two different Errors (SetupError, InterfaceTestError) to achieve the same goal